---
title: "5_prediction_accuracy"
author: "SDS"
date: "Last compiled on `r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    df_print: kable
    code_folding: hide
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding,
  output_dir = paste0(Sys.getenv("RKJCOLLAB"), "/Collabs/metaboxcan/reports")) })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(
  root.dir = paste0(Sys.getenv("RKJCOLLAB"), "/Collabs/metaboxcan"))

library(tidyverse)
library(RSQLite)
library(data.table)
library(purrr)
library(ggpubr)
library(table1)
```

# **METSIM Db Prediction Accuracy in COPDGene**

Prediction accuracy analysis for Haky and Festus using their METSIM metabolite
prediction models in COPDGene. There are 789 metabolites in common between the
METSIM prediction models and the COPDGene metabolite data. There are 957
individuals in COPDGene with both genetic data and metabolite data (from Phase 2
of COPDGene). Prediction accuracy was also calculated separately by sex, which
was requested since METSIM is only male (TODO: confirm?). There are 489 males
and 468 females.

First, residuals of metabolite values were calculated in 3 different ways in
script 4.1_get_metab_residuals.R:

  + 1. age and sex (most closely matches paper draft)
  + 2. age, sex, smoking status, pack years, BMI, and FEV1pp (suggested
      by Katerina)
      + TO NOTE: don't need to adjust for race since subset to NHW

Subset to the 957 individuals that calculating accuracy in before running
model to get residuals here. To do this and keep all covars for option 2,
reduce to 953 individuals. Will run option 1 with all 957 and with 953 to
be comparable. Results for unadjusted metabolite values also shown.

Then, two-sided Pearson correlation prediction accuracy was calculated in
4_calc_prediction_accuracy.R, and details for the steps before that are
available in each script.

In part, Pearson correlation was chosen based on the Predixcan paper by Gamazon
et al (PMID 6258848). From the methods:
"To assess performance, we used the square of the Pearson correlation, R2,
between predicted and observed expression levels."

Spearman correlation was also calculated, but with many warnings due to shared
ties.

# **Load Data**

```{r load_data, message = F, warning = F}

# Predicted Gene Expression
pred <- read.delim(
  "data/metsim_prediction/COPDGene_metaboxcan_predict.txt")

### Unadjusted
# Accuracy
acc_no_adj <- read_tsv(paste0(
  "data/metsim_prediction/COPDGene_metsim_prediction_performance.txt"))

acc_no_adj_m <- read_tsv(paste0(
  "data/metsim_prediction/COPDGene_metsim_prediction_performance_males.txt"))

acc_no_adj_f <- read_tsv(paste0(
  "data/metsim_prediction/COPDGene_metsim_prediction_performance_females.txt"))

# Predicted vs. observed
merge_no_adj <- read_tsv(paste0(
  file="data/metsim_prediction/COPDGene_pred_v_obs_metsim.txt"))

# Metab value
metab_no_adj <- read_csv(
  "raw/metab/COPDGene_P2_LT20miss_knnImp_NoOut_metabs_20211021.csv")

### Simple covar 1 - all individuals
# Accuracy
version <- "age_sex_covar_adj"
acc_c1 <- read_tsv(paste0(
  "data/metsim_prediction/COPDGene_", version, "_metsim_prediction_performance.txt"))

acc_c1_m <- read_tsv(paste0(
  "data/metsim_prediction/COPDGene_", version, "_metsim_prediction_performance_males.txt"))

acc_c1_f <- read_tsv(paste0(
  "data/metsim_prediction/COPDGene_", version, "_metsim_prediction_performance_females.txt"))

# Predicted vs. observed
merge_c1 <- read_tsv(paste0(
  file="data/metsim_prediction/COPDGene_", version, "_pred_v_obs_metsim.txt"))

# Metab value
metab_c1 <- read_tsv(paste0(
  "data/metab/COPDGene_P2_metabs_", version, "_resid.txt"))

### Simple covar 1 - subset of individuals
# Accuracy
version <- "age_sex_covar_subset_adj"
acc_c1_sub <- read_tsv(paste0(
  "data/metsim_prediction/COPDGene_", version, "_metsim_prediction_performance.txt"))

acc_c1_sub_m <- read_tsv(paste0(
  "data/metsim_prediction/COPDGene_", version, "_metsim_prediction_performance_males.txt"))

acc_c1_sub_f <- read_tsv(paste0(
  "data/metsim_prediction/COPDGene_", version, "_metsim_prediction_performance_females.txt"))

# Predicted vs. observed
merge_c1_sub <- read_tsv(paste0(
  file="data/metsim_prediction/COPDGene_", version, "_pred_v_obs_metsim.txt"))

# Metab value
metab_c1_sub <- read_tsv(paste0(
  "data/metab/COPDGene_P2_metabs_", version, "_resid.txt"))

### Full covar 2 - subset of individuals
# Accuracy
version <- "full_covar_subset_adj"
acc_c2 <- read_tsv(paste0(
  "data/metsim_prediction/COPDGene_", version, "_metsim_prediction_performance.txt"))

acc_c2_m <- read_tsv(paste0(
  "data/metsim_prediction/COPDGene_", version, "_metsim_prediction_performance_males.txt"))

acc_c2_f <- read_tsv(paste0(
  "data/metsim_prediction/COPDGene_", version, "_metsim_prediction_performance_females.txt"))

# Predicted vs. observed
merge_c2 <- read_tsv(paste0(
  file="data/metsim_prediction/COPDGene_", version, "_pred_v_obs_metsim.txt"))

# Metab value
metab_c2 <- read_tsv(paste0(
  "data/metab/COPDGene_P2_metabs_", version, "_resid.txt"))

# Metsim db
db <- dbConnect(
  RSQLite::SQLite(),
  "raw/dbs/metsim-invnorm-softimpute.db")
db_ext <- dbReadTable(db, "extra")
dbDisconnect(db)

# Pheno
pheno <- read_tsv(
  "raw/pheno/COPDGene_P1P2P3_Flat_SM_NS_Sep24.txt")
# TO NOTE: this throws warnings, but doesn't affect the column I need (gender)
# 1 = male, 2 = female

# Heritability
herit <- read_csv("data/gcta/merge_constrain_no_covar.csv")

# For mapping of metab_id to CHEM_ID
metab_info <- read_csv(
  "raw/metab/COPDGene_P2_MetaboliteInformation_20211021.csv")
# Mapping:
# metabolite data (including herit data): metab_id --> CHEM_ID
# predicted expression: remove C from IDs, match to CHEM_ID


```


# **Prediction Accuracy Overview**

### No Residuals

**Distribution of Pearson accuracy (R squared, also shown as R2):**

```{r}
# TO NOTE: udpate df "acc" for each section
acc <- acc_no_adj
acc_f <- acc_no_adj_f
acc_m <- acc_no_adj_m
merge <- merge_no_adj
```

```{r, out.width = '50%', fig.show = 'hold', message = F}
ggplot(acc, aes(x = pearson.r2)) +
  geom_histogram() + 
  labs(title = "Pearson R2")

ggplot(acc, aes(x = log(pearson.r2))) +
  geom_histogram() + 
  labs(title = "Log Transform of Pearson R2")

```

**Distribution of Pearson R2 (not transformed):**

```{r}
print(paste0("Min: ", min(acc$pearson.r2, na.rm = T)))
print(paste0("Max: ", max(acc$pearson.r2, na.rm = T)))
quantile(acc$pearson.r2, na.rm = T)
```


**Distribution of Pearson accuracy stratified by sex:**

Correlation here was calculated separately in males and females.

```{r, out.width = '50%', fig.show = 'hold', message = F}
acc_m$sex = "male"
acc_f$sex = "female"
acc_sex <- rbind(acc_m, acc_f)


ggplot(acc_sex, aes(x = pearson.r2, fill = sex)) +
  geom_histogram(position = "identity", alpha = 0.5) + 
  labs(title = "Pearson R2")

ggplot(acc_sex, aes(x = log(pearson.r2), fill = sex)) +
  geom_histogram(position = "identity", alpha = 0.5) + 
  labs(title = "Log Transform of Pearson R2")

```


**How many with Pearson R2 > 0.1?**

```{r}
sum(acc$pearson.r2 > 0.1)
```

**How many significantly correlated (Pearson p-value < 0.05)?**

```{r}
sum(acc$pearson.pval < 0.05)
```


### Residuals, Simple Covar, All 957

**Distribution of Pearson accuracy (R squared, also shown as R2):**

```{r}
# TO NOTE: udpate df "acc" for each section
acc <- acc_c1
acc_f <- acc_c1_f
acc_m <- acc_c1_m
merge <- merge_c1
```


```{r, out.width = '50%', fig.show = 'hold', message = F}
ggplot(acc, aes(x = pearson.r2)) +
  geom_histogram() + 
  labs(title = "Pearson R2")

ggplot(acc, aes(x = log(pearson.r2))) +
  geom_histogram() + 
  labs(title = "Log Transform of Pearson R2")

```

**Distribution of Pearson R2 (not transformed):**

```{r}
print(paste0("Min: ", min(acc$pearson.r2, na.rm = T)))
print(paste0("Max: ", max(acc$pearson.r2, na.rm = T)))
quantile(acc$pearson.r2, na.rm = T)
```


**Distribution of Pearson accuracy stratified by sex:**

Correlation here was calculated separately in males and females.

```{r, out.width = '50%', fig.show = 'hold', message = F}
acc_m$sex = "male"
acc_f$sex = "female"
acc_sex <- rbind(acc_m, acc_f)


ggplot(acc_sex, aes(x = pearson.r2, fill = sex)) +
  geom_histogram(position = "identity", alpha = 0.5) + 
  labs(title = "Pearson R2")

ggplot(acc_sex, aes(x = log(pearson.r2), fill = sex)) +
  geom_histogram(position = "identity", alpha = 0.5) + 
  labs(title = "Log Transform of Pearson R2")

```


**How many with Pearson R2 > 0.1?**

```{r}
sum(acc$pearson.r2 > 0.1)
```

**How many significantly correlated (Pearson p-value < 0.05)?**

```{r}
sum(acc$pearson.pval < 0.05)
```



### Residuals, Simple Covar, Subset 953

**Distribution of Pearson accuracy (R squared, also shown as R2):**

```{r}
# TO NOTE: udpate df "acc" for each section
acc <- acc_c1_sub
acc_f <- acc_c1_sub_f
acc_m <- acc_c1_sub_m
merge <- merge_c1_sub
```


```{r, out.width = '50%', fig.show = 'hold', message = F}
ggplot(acc, aes(x = pearson.r2)) +
  geom_histogram() + 
  labs(title = "Pearson R2")

ggplot(acc, aes(x = log(pearson.r2))) +
  geom_histogram() + 
  labs(title = "Log Transform of Pearson R2")

```

**Distribution of Pearson R2 (not transformed):**

```{r}
print(paste0("Min: ", min(acc$pearson.r2, na.rm = T)))
print(paste0("Max: ", max(acc$pearson.r2, na.rm = T)))
quantile(acc$pearson.r2, na.rm = T)
```


**Distribution of Pearson accuracy stratified by sex:**

Correlation here was calculated separately in males and females.

```{r, out.width = '50%', fig.show = 'hold', message = F}
acc_m$sex = "male"
acc_f$sex = "female"
acc_sex <- rbind(acc_m, acc_f)


ggplot(acc_sex, aes(x = pearson.r2, fill = sex)) +
  geom_histogram(position = "identity", alpha = 0.5) + 
  labs(title = "Pearson R2")

ggplot(acc_sex, aes(x = log(pearson.r2), fill = sex)) +
  geom_histogram(position = "identity", alpha = 0.5) + 
  labs(title = "Log Transform of Pearson R2")

```


**How many with Pearson R2 > 0.1?**

```{r}
sum(acc$pearson.r2 > 0.1)
```

**How many significantly correlated (Pearson p-value < 0.05)?**

```{r}
sum(acc$pearson.pval < 0.05)
```





### Residuals, Full Covar, Subset 953

**Distribution of Pearson accuracy (R squared, also shown as R2):**

```{r}
# TO NOTE: udpate df "acc" for each section
acc <- acc_c2
acc_f <- acc_c2_f
acc_m <- acc_c2_m
merge <- merge_c2
```


```{r, out.width = '50%', fig.show = 'hold', message = F}
ggplot(acc, aes(x = pearson.r2)) +
  geom_histogram() + 
  labs(title = "Pearson R2")

ggplot(acc, aes(x = log(pearson.r2))) +
  geom_histogram() + 
  labs(title = "Log Transform of Pearson R2")

```

**Distribution of Pearson R2 (not transformed):**

```{r}
print(paste0("Min: ", min(acc$pearson.r2, na.rm = T)))
print(paste0("Max: ", max(acc$pearson.r2, na.rm = T)))
quantile(acc$pearson.r2, na.rm = T)
```


**Distribution of Pearson accuracy stratified by sex:**

Correlation here was calculated separately in males and females.

```{r, out.width = '50%', fig.show = 'hold', message = F}
acc_m$sex = "male"
acc_f$sex = "female"
acc_sex <- rbind(acc_m, acc_f)


ggplot(acc_sex, aes(x = pearson.r2, fill = sex)) +
  geom_histogram(position = "identity", alpha = 0.5) + 
  labs(title = "Pearson R2")

ggplot(acc_sex, aes(x = log(pearson.r2), fill = sex)) +
  geom_histogram(position = "identity", alpha = 0.5) + 
  labs(title = "Log Transform of Pearson R2")

```


**How many with Pearson R2 > 0.1?**

```{r}
sum(acc$pearson.r2 > 0.1)
```

**How many significantly correlated (Pearson p-value < 0.05)?**

```{r}
sum(acc$pearson.pval < 0.05)
```







# **Prediction Accuracy Scatter Plots**

### No Residuals

```{r}
# TO NOTE: udpate df "acc" for each section
acc <- acc_no_adj
acc_f <- acc_no_adj_f
acc_m <- acc_no_adj_m
merge <- merge_no_adj
```

**All metabolites, prediction accuracy R2 vs. METSIM model R2:**

```{r}
# identical(db_ext$gene, db_ext$genename)  # TRUE
db_ext_mod <- db_ext %>%
  dplyr::mutate(metab = gsub("^C", "", gene))
acc_merge <- inner_join(
  acc %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))
# sum(is.na(acc_merge$pearson.r2))  # 0
# acc_merge_mod <- acc_merge %>%
#   dplyr::mutate(pearson.r2 = ifelse(is.na(pearson.r2), -0.1, pearson.r2))

plot_acc <- ggplot(acc_merge, aes(x = pred.perf.R2, y = pearson.r2)) +
  geom_point(alpha = 0.5) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2))) +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal() +
  theme(legend.position = "none")

plot_acc

```

*TO NOTE: Not sure why some of their model performance R2 values are negative?*

**All metabolites, prediction accuracy R2 vs. METSIM model R2, stratified by sex:**

Correlation here was calculated separately in males and females. Combined plot
has two points for each metabolite - one showing the correlation calculated in
males and the other in females. The plots are also shown separated.

```{r}
acc_sex_merge <- inner_join(
  acc_sex %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))

ggplot(acc_sex_merge, aes(x = pred.perf.R2, y = pearson.r2, color = sex)) +
  geom_point(alpha = 0.5) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2))) +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal()

```


```{r out.width = '50%', fig.show = 'hold'}
acc_m_merge <- inner_join(
  acc_m %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))
acc_f_merge <- inner_join(
  acc_f %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))

default_colors <- scales::hue_pal()(2)

ggplot(acc_m_merge, aes(x = pred.perf.R2, y = pearson.r2)) +
  geom_point(alpha = 0.5, color = default_colors[2]) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2)),
    title = "Males") +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal()

ggplot(acc_f_merge, aes(x = pred.perf.R2, y = pearson.r2)) +
  geom_point(alpha = 0.5, color = default_colors[1]) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2)),
    title = "Females") +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal()
```


**List metabolites with the highest prediction accuracy:**

```{r}
acc_merge_show <- acc_merge %>%
  dplyr::select(
    metab, pearson.cor, pearson.r2, t.stat, pearson.pval, pred.perf.R2,
    n.snps.in.model) %>%
  dplyr::arrange(desc(pearson.r2))

# switch p-value column to character for display in report
head(
  acc_merge_show %>% dplyr::mutate(pearson.pval = as.character(pearson.pval)), 10)

acc_merge_show_top <- acc_merge_show %>%
  slice(1:4)

```

**Predicted vs. observed plots for top four highest prediction accuracy:**

```{r, warning = F}

cand_gene_plot <- function(data, gene) {

  data_ex <- data %>%
    dplyr::select(ID, pop, all_of(gene)) %>%
    pivot_wider(names_from = c("pop"), values_from = c(gene))
  
  ggplot(data_ex, aes(x = pred.expr, y = obs.expr)) +
    geom_point() +
    stat_cor(method = "pearson", alternative = "two.sided", cor.coef.name = "R",
             p.digits = 5,
             aes(label = paste(after_stat(rr.label), after_stat(p.label), sep = "~`,`~"))) +
    geom_smooth(method = "lm", se = F, color = "black") +  # linear equivalent to pearson
    theme_minimal() +
    labs(
      x = "Predicted Expression",
      y = "Observed Expression",
      color = "",
      title = paste0("Metab CHEM_ID: ", gene)
    ) +
    theme(legend.position = "bottom")
}

invisible(lapply(acc_merge_show_top$metab, function(m) {
  print(cand_gene_plot(merge, m)) }))

```

*TO NOTE: actual p-values are shown in table above, not sure why plot rounding*
*is cuttting all off at 2.2E-16.*



### Residuals, Simple Covar, All 957

```{r}
# TO NOTE: udpate df "acc" for each section
acc <- acc_c1
acc_f <- acc_c1_f
acc_m <- acc_c1_m
merge <- merge_c1
```

**All metabolites, prediction accuracy R2 vs. METSIM model R2:**

```{r}
# identical(db_ext$gene, db_ext$genename)  # TRUE
db_ext_mod <- db_ext %>%
  dplyr::mutate(metab = gsub("^C", "", gene))
acc_merge <- inner_join(
  acc %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))
# sum(is.na(acc_merge$pearson.r2))  # 0
# acc_merge_mod <- acc_merge %>%
#   dplyr::mutate(pearson.r2 = ifelse(is.na(pearson.r2), -0.1, pearson.r2))

plot_acc <- ggplot(acc_merge, aes(x = pred.perf.R2, y = pearson.r2)) +
  geom_point(alpha = 0.5) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2))) +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal() +
  theme(legend.position = "none")

plot_acc

```

*TO NOTE: Not sure why some of their model performance R2 values are negative?*

**All metabolites, prediction accuracy R2 vs. METSIM model R2, stratified by sex:**

Correlation here was calculated separately in males and females. Combined plot
has two points for each metabolite - one showing the correlation calculated in
males and the other in females. The plots are also shown separated.

```{r}
acc_sex_merge <- inner_join(
  acc_sex %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))

ggplot(acc_sex_merge, aes(x = pred.perf.R2, y = pearson.r2, color = sex)) +
  geom_point(alpha = 0.5) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2))) +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal()

```


```{r out.width = '50%', fig.show = 'hold'}
acc_m_merge <- inner_join(
  acc_m %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))
acc_f_merge <- inner_join(
  acc_f %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))

default_colors <- scales::hue_pal()(2)

ggplot(acc_m_merge, aes(x = pred.perf.R2, y = pearson.r2)) +
  geom_point(alpha = 0.5, color = default_colors[2]) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2)),
    title = "Males") +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal()

ggplot(acc_f_merge, aes(x = pred.perf.R2, y = pearson.r2)) +
  geom_point(alpha = 0.5, color = default_colors[1]) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2)),
    title = "Females") +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal()
```


**List metabolites with the highest prediction accuracy:**

```{r}
acc_merge_show <- acc_merge %>%
  dplyr::select(
    metab, pearson.cor, pearson.r2, t.stat, pearson.pval, pred.perf.R2,
    n.snps.in.model) %>%
  dplyr::arrange(desc(pearson.r2))

# switch p-value column to character for display in report
head(
  acc_merge_show %>% dplyr::mutate(pearson.pval = as.character(pearson.pval)), 10)

acc_merge_show_top <- acc_merge_show %>%
  slice(1:4)

```


### Residuals, Simple Covar, Subset 953

```{r}
# TO NOTE: udpate df "acc" for each section
acc <- acc_c1_sub
acc_f <- acc_c1_sub_f
acc_m <- acc_c1_sub_m
merge <- merge_c1_sub
```

**All metabolites, prediction accuracy R2 vs. METSIM model R2:**

```{r}
# identical(db_ext$gene, db_ext$genename)  # TRUE
db_ext_mod <- db_ext %>%
  dplyr::mutate(metab = gsub("^C", "", gene))
acc_merge <- inner_join(
  acc %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))
# sum(is.na(acc_merge$pearson.r2))  # 0
# acc_merge_mod <- acc_merge %>%
#   dplyr::mutate(pearson.r2 = ifelse(is.na(pearson.r2), -0.1, pearson.r2))

plot_acc <- ggplot(acc_merge, aes(x = pred.perf.R2, y = pearson.r2)) +
  geom_point(alpha = 0.5) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2))) +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal() +
  theme(legend.position = "none")

plot_acc

```

*TO NOTE: Not sure why some of their model performance R2 values are negative?*

**All metabolites, prediction accuracy R2 vs. METSIM model R2, stratified by sex:**

Correlation here was calculated separately in males and females. Combined plot
has two points for each metabolite - one showing the correlation calculated in
males and the other in females. The plots are also shown separated.

```{r}
acc_sex_merge <- inner_join(
  acc_sex %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))

ggplot(acc_sex_merge, aes(x = pred.perf.R2, y = pearson.r2, color = sex)) +
  geom_point(alpha = 0.5) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2))) +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal()

```


```{r out.width = '50%', fig.show = 'hold'}
acc_m_merge <- inner_join(
  acc_m %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))
acc_f_merge <- inner_join(
  acc_f %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))

default_colors <- scales::hue_pal()(2)

ggplot(acc_m_merge, aes(x = pred.perf.R2, y = pearson.r2)) +
  geom_point(alpha = 0.5, color = default_colors[2]) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2)),
    title = "Males") +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal()

ggplot(acc_f_merge, aes(x = pred.perf.R2, y = pearson.r2)) +
  geom_point(alpha = 0.5, color = default_colors[1]) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2)),
    title = "Females") +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal()
```


**List metabolites with the highest prediction accuracy:**

```{r}
acc_merge_show <- acc_merge %>%
  dplyr::select(
    metab, pearson.cor, pearson.r2, t.stat, pearson.pval, pred.perf.R2,
    n.snps.in.model) %>%
  dplyr::arrange(desc(pearson.r2))

# switch p-value column to character for display in report
head(
  acc_merge_show %>% dplyr::mutate(pearson.pval = as.character(pearson.pval)), 10)

acc_merge_show_top <- acc_merge_show %>%
  slice(1:4)

```


### Residuals, Full Covar, Subset 953

```{r}
# TO NOTE: udpate df "acc" for each section
acc <- acc_c2
acc_f <- acc_c2_f
acc_m <- acc_c2_m
merge <- merge_c2
```

**All metabolites, prediction accuracy R2 vs. METSIM model R2:**

```{r}
# identical(db_ext$gene, db_ext$genename)  # TRUE
db_ext_mod <- db_ext %>%
  dplyr::mutate(metab = gsub("^C", "", gene))
acc_merge <- inner_join(
  acc %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))
# sum(is.na(acc_merge$pearson.r2))  # 0
# acc_merge_mod <- acc_merge %>%
#   dplyr::mutate(pearson.r2 = ifelse(is.na(pearson.r2), -0.1, pearson.r2))

plot_acc <- ggplot(acc_merge, aes(x = pred.perf.R2, y = pearson.r2)) +
  geom_point(alpha = 0.5) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2))) +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal() +
  theme(legend.position = "none")

plot_acc

```

*TO NOTE: Not sure why some of their model performance R2 values are negative?*

**All metabolites, prediction accuracy R2 vs. METSIM model R2, stratified by sex:**

Correlation here was calculated separately in males and females. Combined plot
has two points for each metabolite - one showing the correlation calculated in
males and the other in females. The plots are also shown separated.

```{r}
acc_sex_merge <- inner_join(
  acc_sex %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))

ggplot(acc_sex_merge, aes(x = pred.perf.R2, y = pearson.r2, color = sex)) +
  geom_point(alpha = 0.5) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2))) +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal()

```


```{r out.width = '50%', fig.show = 'hold'}
acc_m_merge <- inner_join(
  acc_m %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))
acc_f_merge <- inner_join(
  acc_f %>% dplyr::mutate(metab = as.character(metab)),
  db_ext_mod,
  by = c("metab"))

default_colors <- scales::hue_pal()(2)

ggplot(acc_m_merge, aes(x = pred.perf.R2, y = pearson.r2)) +
  geom_point(alpha = 0.5, color = default_colors[2]) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2)),
    title = "Males") +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal()

ggplot(acc_f_merge, aes(x = pred.perf.R2, y = pearson.r2)) +
  geom_point(alpha = 0.5, color = default_colors[1]) +
  coord_fixed() +
  labs(
    x = bquote("Model Performance"~(R^2)),
    y = bquote("Prediction Accuracy"~(R^2)),
    title = "Females") +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min(acc_merge$pred.perf.R2),1)) +
  theme_minimal()
```


**List metabolites with the highest prediction accuracy:**

```{r}
acc_merge_show <- acc_merge %>%
  dplyr::select(
    metab, pearson.cor, pearson.r2, t.stat, pearson.pval, pred.perf.R2,
    n.snps.in.model) %>%
  dplyr::arrange(desc(pearson.r2))

# switch p-value column to character for display in report
head(
  acc_merge_show %>% dplyr::mutate(pearson.pval = as.character(pearson.pval)), 10)

acc_merge_show_top <- acc_merge_show %>%
  slice(1:4)

```



### Compare Accuracy Across Versions

```{r}
# For some reason pivot_wider not working (duplicated values), so instead
# combining by making new columns
acc_c1_comb <- acc_c1 %>%
  dplyr::mutate(covar_short = pearson.r2) %>%
  dplyr::select(metab, covar_short)
acc_c1_sub_comb <- acc_c1_sub %>%
  dplyr::mutate(covar_short_subset = pearson.r2) %>%
  dplyr::select(metab, covar_short_subset)
acc_c2_comb <- acc_c2 %>%
  dplyr::mutate(covar_full = pearson.r2) %>%
  dplyr::select(metab, covar_full)
acc_no_adj_comb <- acc_no_adj %>%
  dplyr::mutate(unadj = pearson.r2) %>%
  dplyr::select(metab, unadj)

acc_comb_t1 <- inner_join(acc_c1_comb, acc_c1_sub_comb, by = c("metab"))
acc_comb_t2 <- inner_join(acc_c2_comb, acc_no_adj_comb, by = c("metab"))
acc_comb <- inner_join(acc_comb_t1, acc_comb_t2, by = c("metab"))

```

**Each version vs. unadjusted:**

```{r }
ggplot(acc_comb, aes(x = unadj, y = covar_short)) +
  geom_point(alpha = 0.5) +
  coord_fixed() +
  labs(title = "Pearson R2") +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(0,1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(0,1)) +
  theme_minimal() +
  theme(legend.position = "none")

ggplot(acc_comb, aes(x = unadj, y = covar_short_subset)) +
  geom_point(alpha = 0.5) +
  coord_fixed() +
  labs(title = "Pearson R2") +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(0,1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(0,1)) +
  theme_minimal() +
  theme(legend.position = "none")

ggplot(acc_comb, aes(x = unadj, y = covar_full)) +
  geom_point(alpha = 0.5) +
  coord_fixed() +
  labs(title = "Pearson R2") +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(0,1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(0,1)) +
  theme_minimal() +
  theme(legend.position = "none")



```

**Residual versions with each other:**

```{r}
ggplot(acc_comb, aes(x = covar_short, y = covar_short_subset)) +
  geom_point(alpha = 0.5) +
  coord_fixed() +
  labs(title = "Pearson R2") +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(0,1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(0,1)) +
  theme_minimal() +
  theme(legend.position = "none")

ggplot(acc_comb, aes(x = covar_short, y = covar_full)) +
  geom_point(alpha = 0.5) +
  coord_fixed() +
  labs(title = "Pearson R2") +
  scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(0,1)) +
  scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(0,1)) +
  theme_minimal() +
  theme(legend.position = "none")
```


All are VERY similar!

# **For Publication**

Matching draft paper supplemental figure 2 - box plots of R and scatter plot
of performance in females vs. males.

### **Table 1**

For the 953 individuals with full covariate adjustment.

Smoking status = 0=Never-smoked | 1=Former smoker | 2=Current smoker

```{r}
c2_ids <- unique(merge_c2$ID)
pheno_c2 <- pheno %>%
  dplyr::filter(sid %in% c2_ids)

table1( ~ Age_P2 + as.factor(race) + as.factor(smoking_status_P2) +
          ATS_PackYears_P2 + BMI_P2 + Pred_FEV1_P2 | as.factor(gender),
        data = pheno_c2)

table1( ~ as.factor(gender) | as.factor(race), data = pheno_c2)
```


### **Pearson**

For now, doing with each of the residual versions since all very similar.

```{r}
plot_pub <- function(data_f, data_m, title) {
  data_f$sex <- "Female"
  data_m$sex <- "Male"
  data_lg <- rbind(data_f, data_m)
  
  data_wd <- inner_join(
    data_f %>% dplyr::select(metab, pearson.cor),
    data_m %>% dplyr::select(metab, pearson.cor),
    by = c("metab"), suffix = c("_f", "_m"))
  
  # coord_fixed ratio is y/x
  ratio <- 2 / (range(data_lg$pearson.cor)[2] - range(data_lg$pearson.cor)[1])
  boxplot <- ggplot(data_lg, aes(x = sex, y = pearson.cor)) +
    coord_fixed(ratio) +
    geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) +
    stat_compare_means(method = "t.test", label = "p.format",
                       label.x = 1.5, label.y = 0.65, hjust = 0.6) +
    labs(x = "Sex",
         y = "Pearson Correlation") +
    theme_bw()
  boxplot
  
  min_r <- min(data_lg$pearson.cor)
  max_r <- max(data_lg$pearson.cor)
  scatter <- ggplot(data_wd, aes(x = pearson.cor_m, y = pearson.cor_f)) +
    geom_point(alpha = 0.5) +
    coord_fixed() +
    labs(
      x = "Males",
      y = "Females") +
    scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min_r, max_r)) +
    scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min_r, max_r)) +
    theme_bw() +
    theme(legend.position = "none") +
    stat_cor(method = "pearson", label.x = 0.35, label.y = 0.65, hjust = 0.5)
  
  plot <- ggarrange(
    boxplot, scatter, nrow = 1, ncol = 2, align = "hv")
  plot <- annotate_figure(
    plot, fig.lab = title, fig.lab.pos = "top.left", fig.lab.face = "bold")
  
  return(plot)
  
}
```

```{r}
plot_pub(acc_c1_f, acc_c1_m, "1. Covar Only Age and Sex, All 957")
plot_pub(acc_c2_f, acc_c2_m, "2. Covar Full, Subset 953")
plot_pub(acc_c1_sub_f, acc_c1_sub_m, "3. Covar Only Age and Sex, Subset 953")
```


**Decision:** for publication, use the full covariate version in 953
individuals, think best to adjsut for all covariates suggested by Katerina
since COPDGene is a unique cohort. Written out for publication at end.

### **Spearman**

```{r}
plot_pub_spearman <- function(data_f, data_m, title) {
  data_f$sex <- "Female"
  data_m$sex <- "Male"
  data_lg <- rbind(data_f, data_m)
  
  data_wd <- inner_join(
    data_f %>% dplyr::select(metab, spearman.cor),
    data_m %>% dplyr::select(metab, spearman.cor),
    by = c("metab"), suffix = c("_f", "_m"))
  
  # coord_fixed ratio is y/x
  ratio <- 2 / (range(data_lg$spearman.cor)[2] - range(data_lg$spearman.cor)[1])
  boxplot <- ggplot(data_lg, aes(x = sex, y = spearman.cor)) +
    coord_fixed(ratio) +
    geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) +
    stat_compare_means(method = "t.test", label = "p.format",
                       label.x = 1.5, label.y = 0.65, hjust = 0.6) +
    labs(x = "Sex",
         y = "Spearman Correlation") +
    theme_bw()
  boxplot
  
  min_r <- min(data_lg$spearman.cor)
  max_r <- max(data_lg$spearman.cor)
  scatter <- ggplot(data_wd, aes(x = spearman.cor_m, y = spearman.cor_f)) +
    geom_point(alpha = 0.5) +
    coord_fixed() +
    labs(
      x = "Males",
      y = "Females") +
    scale_x_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min_r, max_r)) +
    scale_y_continuous(breaks = seq(0, 1, by = 0.3), limits = c(min_r, max_r)) +
    theme_bw() +
    theme(legend.position = "none") +
    stat_cor(method = "spearman", label.x = 0.35, label.y = 0.65, hjust = 0.5)
  
  plot <- ggarrange(
    boxplot, scatter, nrow = 1, ncol = 2, align = "hv")
  plot <- annotate_figure(
    plot, fig.lab = title, fig.lab.pos = "top.left", fig.lab.face = "bold")
  
  return(plot)
  
}
```

```{r}
plot_pub_spearman(acc_c2_f, acc_c2_m, "2. Covar Full, Subset 953")
```


# **Heritability Results - Unadjusted Only**

Currently, heritability was calculated for the observed/actual metabolomics
data (not predicted) in unrelateds (N=953, 4 removed) based on the
GCTA-calculated GRM and a cutoff of 0.05. Constrained REML analysis was then
run in GCTA without any covariates.

*TODO: update this once talk with Haky about methods used in paper.*

**Distribution of heritability:**

```{r}
ggplot(herit, aes(`V(G)/Vp`)) +
  geom_histogram(alpha = 0.5) +
  ggtitle("Heritability") +
  xlim(-0.1, 1)
```

**Distribution of p-value vs. heritability:**

Line shown at a p-value of 0.05.

```{r}
ggplot(herit, aes(x=`V(G)/Vp`, y=-log10(Pval))) +
  geom_point(alpha = 0.5) +
  ggtitle("Pval vs. Heritability") +
  xlim(-0.1,1) +
  ylim(-1,10) +
  geom_hline(yintercept = -log10(0.05))
```


**How many with Pvalue < 0.05?**

```{r}
sum(herit$Pval < 0.05)
```

**How many with heritability > 0.5?**

```{r}
sum(herit$`V(G)/Vp` > 0.5)
```

**How many Pvalue < 0.05 and heritability > 0.5?**

```{r}
herit_sig <- herit %>% dplyr::filter(Pval < 0.05)
sum(herit_sig$`V(G)/Vp` > 0.5)
```





# **Write Out**

```{r}
pub <- plot_pub(acc_c2_f, acc_c2_m, "")

ggsave("dissemination/copdgene_metaboxcan_plot_for_pub.png", pub,
       units = "in", width = 6, height = 3.5)

pub_spearman <- plot_pub_spearman(acc_c2_f, acc_c2_m, "")

ggsave("dissemination/copdgene_metaboxcan_plot_for_pub_spearman.png", pub_spearman,
       units = "in", width = 6, height = 3.5)
```


# **Session Info**

```{r session_info}
sessionInfo()
```

